diff --git a/node_modules/s3rver/lib/controllers/object.js b/node_modules/s3rver/lib/controllers/object.js
index e452c9d..55e5335 100644
--- a/node_modules/s3rver/lib/controllers/object.js
+++ b/node_modules/s3rver/lib/controllers/object.js
@@ -1,6 +1,6 @@
 'use strict';
 
-const Busboy = require('busboy');
+const busboyCons = require('busboy');
 const crypto = require('crypto');
 const xmlParser = require('fast-xml-parser');
 const he = require('he');
@@ -281,7 +281,7 @@ exports.postObject = async function postObject(ctx) {
     let successStatus = 204;
     let fileCount = 0;
 
-    const busboy = new Busboy({ headers: ctx.headers });
+    var busboy = busboyCons({ headers: ctx.headers });
     const finishedPromise = once(busboy, 'finish');
     const object = await Promise.race([
       finishedPromise.then(() => null),
