// @serverless/platform-sdk/dist/apps/createApp.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

var fetch = require('isomorphic-fetch');
var platformConfig = require('../config');
var currentVersion = require('../../package.json').version;

var slugifyApp = function slugifyApp(text) {
  return text.toString().toLowerCase().replace(/[^a-z0-9]+/g, '-') // Replace any run of disallowed chars with a hyphen
  .replace(/^-+/, '') // remove leading hyphens
  .replace(/-+$/, '');
}; // remove trailing hyphens

var createApp = function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(data) {
    var body, response, text;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            body = JSON.stringify({
              tenantName: data.tenant,
              appName: slugifyApp(data.app),
              title: data.app
            });
            _context.next = 3;
            return fetch(`${platformConfig.backendUrl}tenants/${data.tenant}/applications`, {
              method: 'POST',
              body,
              headers: {
                'Content-Type': 'application/json',
                'x-platform-version': currentVersion,
                Authorization: `bearer ${data.token}`
              }
            });

          case 3:
            response = _context.sent;

            if (response.ok) {
              _context.next = 9;
              break;
            }

            _context.next = 7;
            return response.text();

          case 7:
            text = _context.sent;
            throw new Error(text);

          case 9:
            return _context.abrupt('return', response.json());

          case 10:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, undefined);
  }));

  return function createApp(_x) {
    return _ref.apply(this, arguments);
  };
}();

exports.default = createApp;
